Update installation instructio for JAX and add some dependencies.

nouiz opened the pull request.

ptrendx commented on README.rst:
```diff
@@ -131,6 +131,12 @@ Transformer Engine comes preinstalled in the pyTorch container on
 From source
 ^^^^^^^^^^^
 
+For JAX, you need to install those dependencies first. (We are working to make this automatic):
```
I don't think we need to say that we work to make this automatic - at some point we will just drop this section :-).

ptrendx commented on setup.py:
```diff
@@ -161,11 +161,16 @@ def install_requires():
 
+    def install_requires():
+        # TODO: need to find a way to instal directly pybind11 and ninja
+        return ['pydantic', 'cmake']
```
Do we need cmake from pip?

ptrendx replied:
Also, do we actually need pydantic? Its functionality does not seem necessary for supporting TE in JAX?

nouiz replied:
cmake could come from the system. But having it installed from python simplify end user experience.

I think pydantic is needed due to this code:
```
$ git grep pydantic
setup.py:        return ['pydantic', 'cmake']
transformer_engine/common/recipe.py:from pydantic.dataclasses import dataclass
```

ptrendx replied:
@ksivaman Since pydantic is used in common, that should actually be a dependency for all frameworks, right?

timmoon10 replied:
If we're getting CMake from pip, I wonder if that gives us a nicer way to get the CMake executable.

ptrendx commented on setup.py:
```diff
@@ -195,6 +200,11 @@ def run(self, extensions):
+    try:
+        import pybind11
+    except ImportError:
+        print("ERROR: pybind11 not present. Can't install JAX backend without it. You can install it with `pip install pybind11`.")
```
```suggestion
        print("ERROR: pybind11 not present. You can install it with `pip install pybind11`.")
```
Also, is that the right pythonic way? Shouldn't the error message live in the exception instead of a random print?

ksivaman replied:
Yeah I agree, instead of doing this print statement followed by `raise`, a meaningful error with a message must be displayed. In this specific case I'd argue that any exception handling is not needed and python's stack trace for the `ImportError` in case pybind11 is not installed would be sufficient.

nouiz replied:
Good idea. Only doing `import pybind11` now.

nouiz commented:
> Could you sign-off your commits using the instructions [here](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)? @nouiz

Done.

ptrendx commented on docs/installation.rst:
```diff
@@ -30,8 +30,9 @@ Additional Prerequisites
-3. `Ninja <https://ninja-build.org/>`__
+2. `Ninja <https://ninja-build.org/>`__
+3. [For pyTorch support] `pyTorch <https://pytorch.org/>`__ with GPU support
+3. [For JAX support] `JAX <https://github.com/google/jax/>`__ with GPU support
```
Since we know that not every version of JAX is going to work, could we add the specific version number here?

nouiz replied:
Done.

nouiz commented on setup.py:
```diff
@@ -161,11 +161,16 @@ def install_requires():
 
+    def install_requires():
+        # TODO: need to find a way to instal directly pybind11 and ninja
+        return ['cmake']
```
I added them as te[test] target.

ptrendx replied:
Unit tests' requirements are their own thing and should not be installed for every user.

nouiz replied:
This is what I did.

`pip install te` won't install them.
`pip install te[test]` will install them.

timmoon10 replied:
I've been able to install JAX in the TensorFlow container, but I get runtime errors in TE if Flax isn't installed as well.

nouiz commented on setup.py:
```diff
@@ -257,10 +264,16 @@ def build_extensions(self) -> None:
+        except ImportError:
+            pass
+        else:
+            cmake_args.append("-GNinja")
```
I updated the PR to make it clear that ninja is optional.

nouiz commented:
@timmoon10 @ptrendx I did all the requested changes.
I didn't made the cmake packages a pytorch/tf dependency. I wasn't sure you wanted to change pytorch dependencies.
This can be done in a follow up PR if you want to do that.

ksivaman commented:
Could you sign-off your commits using the instructions [here](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)? @nouiz

Please resolve the conflicts too, thanks!

nouiz commented:
I rebased, squashed all the commits and signed it.
I did some fixes due to the rebase conflict. So it can be good that you look at the full diff again.

I didn't make cmake a pip dependency for all frameworks, but I think we should do that in a follow up PR while testing it for all FW. This will simplify the installation instruction.

ptrendx commented on README.rst:
```diff
@@ -162,32 +162,28 @@ Transformer Engine comes preinstalled in the pyTorch container on
 
 .. code-block:: bash
 
-  apt-get install ninja-build pybind11-dev
```
Why is pybind11-dev package not working?

nouiz replied:
It is working. It is just that pip packges works for more OS.
So this is a more general solution.
rpm based OS will work with the pip package.

ptrendx commented:
/te-ci

nouiz commented:
/te-ci

ksivaman commented on setup.py:
```diff
@@ -399,5 +411,8 @@ def get_outputs(self):
     ext_modules=ext_modules,
     cmdclass={"build_ext": TEBuildExtension},
     install_requires=dlfw_install_requires,
+    extras_require={
+        'test': ['pytest',
+                 'tensorflow_datasets'],},
```
Note: `onnxruntime` is needed for the PyTorch tests.

nouiz replied:
I'm not sure how to handle this.
I think adding it for all FW isn't great.
I'm creating a "test_pytorch" extra.

ksivaman commented on README.rst:
```diff
@@ -162,32 +162,28 @@ Transformer Engine comes preinstalled in the pyTorch container on
 ^^^^^^^^^^^
 
-First, install the prequisites.
+For JAX and Tensorflow, you need to install that dependency first:
```
```suggestion
For JAX and Tensorflow, pybind11 must be installed:
```

ksivaman commented on README.rst:
```diff
@@ -162,32 +162,28 @@ Transformer Engine comes preinstalled in the pyTorch container on
+  pip install pybind11
 
-Clone the repository and inside it type:
+Then, you can install this optional dependency:
```
```suggestion
Then, the following optional dependencies can be installed:
```

ksivaman replied:
Stating the purpose of these dependencies would be useful, since they are 'optional'.

ksivaman commented on README.rst:
```diff
@@ -162,32 +162,28 @@ Transformer Engine comes preinstalled in the pyTorch container on
+  pip install ninja
 
-You can also specify which framework bindings to build. The default is pytorch only.
+Clone the repository and inside it type one of those commands:
```
```suggestion
Install TE (optionally specifying the framework):
```

nouiz replied:
I added the clone command line.

ptrendx replied:
Then "type one of those commands" is even more ambiguous. I agree with Kithi's suggestion.

ksivaman commented on README.rst:
```diff
@@ -162,32 +162,28 @@ Transformer Engine comes preinstalled in the pyTorch container on
 
-  # Build with TensorFlow bindings
-  NVTE_FRAMEWORK=tensorflow pip install .
+  NVTE_FRAMEWORK=all pip install .        # Building with all frameworks.
+  NVTE_FRAMEWORK=pytorch pip install .    # Building with pyTorch only.
+  NVTE_FRAMEWORK=jax pip install .        # Building with JAX only.
+  NVTE_FRAMEWORK=tensorflow pip install . # Building with Tensorflow only.
```
```suggestion
  NVTE_FRAMEWORK=all pip install .        # Build TE for all supported frameworks.
  NVTE_FRAMEWORK=pytorch pip install .    # Build TE for PyTorch only.
  NVTE_FRAMEWORK=jax pip install .        # Build TE for JAX only.
  NVTE_FRAMEWORK=tensorflow pip install . # Build TE for TensorFlow only.
```

ksivaman commented on README.rst:
```diff
@@ -162,32 +162,28 @@ Transformer Engine comes preinstalled in the pyTorch container on
-
-  # Build with all bindings (Pytorch, TF, Jax)
-  NVTE_FRAMEWORK=all pip install .
+The default is pyTorch only.
```
```suggestion
If the framework is not explicitly specified, TE will be built for PyTorch only.
```

ksivaman commented on setup.py:
```diff
@@ -160,23 +160,24 @@ def run(self, extensions):
-    @staticmethod
     def install_requires():
-        return ["pydantic",]
+        # TODO: need to find a way to instal directly pybind11 and ninja
```
```suggestion
        # TODO: find a way to install pybind11 and ninja directly.
```

ksivaman commented on setup.py:
```diff
@@ -207,11 +208,16 @@ def install_requires():
 
 if framework in ("all", "jax"):
     dlfw_builder_funcs.append(JaxBuilder)
+    # Trigger a better error when pybind11 isn't present.
+    # Sadly, if pybind11 was installed with `apt -y install pybind11-dev`
+    # This doesn't install a python packages. So the line bellow is too strict.
+    # When it fail, we need to detect if cmake will find pybind11.
+    # import pybind11
```
This is not clear, what is the goal of this comment?

ksivaman replied:
Is it to answer the question, "Why we're getting pybind11 via cmake?"

timmoon10 commented on docs/installation.rst:
```diff
@@ -29,9 +29,11 @@ pip - from GitHub
-1. `CMake <https://cmake.org/>`__ version 3.18 or later
-2. `pyTorch <https://pytorch.org/>`__ with GPU support
-3. `Ninja <https://ninja-build.org/>`__
+1. [Not needed for JAX] `CMake <https://cmake.org/>`__ version 3.18 or later.
+2. [Optional] `Ninja <https://ninja-build.org/>`__: `pip install ninja`.
+3. [For pyTorch support] `pyTorch <https://pytorch.org/>`__ with GPU support.
+3. [For JAX support] `JAX <https://github.com/google/jax/>`__ with GPU support, version >= 0.4.7.
+3. [For JAX support] `pybind11`: `pip install pybind11`.
```
```suggestion
1. `CMake <https://cmake.org/>`__ version 3.18 or later.
2. `pybind11 <https://github.com/pybind/pybind11>`__ : `pip install pybind11`.
3. [Optional] `Ninja <https://ninja-build.org/>`__: `pip install ninja`.
4. [For PyTorch support] `PyTorch <https://pytorch.org/>`__ with GPU support.
5. [For JAX support] `JAX <https://github.com/google/jax/>`__ with GPU support, version >= 0.4.7.
6. [For TensorFlow support] `TensorFlow <https://www.tensorflow.org/>`__ with GPU support.
```
CMake and pybind11 are used to build the core library, so I don't think they should be treated as optional . The reason we don't need to install pybind11 in the PyTorch build is because it's a PyTorch dependency.

nouiz replied:
This isn't an optional dependency. There isn't an optional section.
To prevent confusion, I just added a commit that move the optional one at the end of the list.

nouiz replied:
updated to reflect that.

timmoon10 reviewed (approved):
LGTM

nouiz commented:
@ptrendx I accepted the suggested modification. But I don't see it anymore, so I can't comment on it.
